### PR TITLE
Assignment6 lambda expressions Matthew Adams, Rachael DeBoer

### PR DESCRIPTION
**Pull request by MattA41**
@MattA41 and @rkdeboer

### CODE REVIEW

**Comment by Aramirez61** [comment]
I'll review this.

File: Calculate/Program.cs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    {
+        WriteLine ??= Console.WriteLine;
+        ReadLine ??= Console.ReadLine;
+    }
+    public static void Main()
+    {
+
+        string? expression = "";
+        Program program = new();
+        do
```
Comment (Aramirez61) [inline Calculate/Program.cs:25]: Here, to follow the requirement "Implement the Program class to instantiate the calculator and invoke it based on user input from the console. ❌"
you should instantiate a calculator object like Calculator calculator = new();

File: Calculate/Program.cs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        {
+            program.WriteLine("Please Enter Your Expression: ");
+            expression = program.ReadLine();
+            Calculator.TryCalculate(expression!, out int result);
```
Comment (Aramirez61) [inline Calculate/Program.cs:29]: This line would also change according to what you name your calculator object when you call the static method.

File: CalculateTests/ProgramTests.cs [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+                if (text != null)
+                {
+                    Assert.Equal(expectedOutput, text);
+                }
+                else
+                {
+                    Assert.True(false, "The output should not be null.");
+                }
+            },
+            ReadLine = () =>
```
Comment (Aramirez61) [inline CalculateTests/ProgramTests.cs:27]: Good job using the WriteLine and ReadLine delegates to assert your output is correct

File: Calculate/Calculator.cs [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    public static int Subtract(int one, int two)
+    {
+        return one - two;
+    }
+    public static int Multiple(int one, int two)
+    {
+        return one * two;
+    }
+
+    public static int Divide(int one, int two)
```
Comment (Aramirez61) [inline Calculate/Calculator.cs:32]: You may want to consider using doubles instead of ints to handle division correctly. Also consider the case of dividing 0.

**Review by Aramirez61 — COMMENTED** [review]
Overall, good job following the assignment requirements, just need some small adjustments for instantiating the calculator in your main, and handling edge cases for your divide method. Also make sure to add a directory build props file.

Create a Console project called "Calculate.". ✔
Define a Program Class
Define two init-only setter properties, WriteLine and ReadLine, that contain delegates for writing a line of text and reading a line of text respectively ✔
Write a test that sets these properties at construction time and then invokes the properties and verifies the expected behavior occurs. ✔
Set the default behavior for the WriteLine and ReadLine properties to invoke System.Console versions of the methods and add an empty default constructor. ✔
Define a Calculator class ✔
Define static Add, Subtract, Multiple, and Divide methods that have two parameters and return a third parameter. ✔
Define a read-only property, MathematicalOperations, of type System.Collections.Generics.IReadOnlyDictionary<TKey,TValue> that:
is initialized to a System.Collections.Generics.Dictionary<<TKey,TValue> instance that. ✔
Uses char for the key corresponding to the operators +, -, *, and /. ✔
Has values that correspond with the Add, Subtract, Multiple, and Divide methods. ✔
Implement a TryCalculate method following "TryParse" pattern ✔
Valid calculation expressions include such strings as "3 + 4", "42 - 2", etc. ✔
If there is no whitespace around the operator, you can assume the calculation is invalid and return false. Similarly if the operands are not integers. ✔
Use string.Split(), pattern matching, logical and operators to parse the string in their entirety ✔
Index into the MathematicalOperations method using the operator parsed during pattern matching to find the corresponding implementation and invoke it. ✔
Implement the Program class to instantiate the calculator and invoke it based on user input from the console. ❌
(In the program class you need to instantiate a new calculator object)

Be sure to use the WriteLine/ReadLine properties on Program for testing the input and output of your program. ✔


Fundamentals
Place all shared project properties into a Directory.Build.Props file.
(Make sure to add this)

Place all shared project items into a Directory.Build.targets file.
nullable reference types is enabled ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
and enabled .NET analyzers for both projects ✔
For this assignment, always use Assert.AreEqual<T>() (the generic version) ✔
All of the above should be unit tested ✔
Choose simplicity over complexity ✔

**Comment by SergeiAshikhmin** [comment]
I will review this

File: CalculateTests/CalculateTests.csproj [hunk 5]
```diff
@@ -0,0 +1,29 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>net8.0</TargetFramework>
+    <ImplicitUsings>enable</ImplicitUsings>
+    <Nullable>enable</Nullable>
+
```
Comment (SergeiAshikhmin) [inline CalculateTests/CalculateTests.csproj:7]: Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ❌

File: Calculate/Calculate.csproj [hunk 6]
```diff
@@ -0,0 +1,16 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFramework>net8.0</TargetFramework>
```
Comment (SergeiAshikhmin) [inline Calculate/Calculate.csproj:5]: Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ❌

File: CalculateTests/ProgramTests.cs (hunk 3 above)
Comment (SergeiAshikhmin) [reply CalculateTests/ProgramTests.cs:27]: I second this suggestion.
Implement the Program class to instantiate the calculator and invoke it based on user input from the console. ❌

**Review by SergeiAshikhmin — COMMENTED** [review]
Good job, only a few details are missing in fundamentals and the Calculator instantiation.

Instructions
Create a Console project called "Calculate.". ✔
Define a Program Class
Define two init-only setter properties, WriteLine and ReadLine, that contain delegates for writing a line of text and reading a line of text respectively ✔
Write a test that sets these properties at construction time and then invokes the properties and verifies the expected behavior occurs. ✔
Set the default behavior for the WriteLine and ReadLine properties to invoke System.Console versions of the methods and add an empty default constructor. ✔
Define a Calculator class ✔
Define static Add, Subtract, Multiple, and Divide methods that have two parameters and return a third parameter. ✔
Define a read-only property, MathematicalOperations, of type System.Collections.Generics.IReadOnlyDictionary<TKey,TValue> that:
is initialized to a System.Collections.Generics.Dictionary<<TKey,TValue> instance that. ✔
Uses char for the key corresponding to the operators +, -, *, and /. ✔
Has values that correspond with the Add, Subtract, Multiple, and Divide methods. ✔
Implement a TryCalculate method following "TryParse" pattern ✔
Valid calculation expressions include such strings as "3 + 4", "42 - 2", etc. ✔
If there is no whitespace around the operator, you can assume the calculation is invalid and return false. Similarly if the operands are not integers. ✔
Use string.Split(), pattern matching, logical and operators to parse the string in their entirety ✔
Index into the MathematicalOperations method using the operator parsed during pattern matching to find the corresponding implementation and invoke it. ✔
Implement the Program class to instantiate the calculator and invoke it based on user input from the console. ❌
Be sure to use the WriteLine/ReadLine properties on Program for testing the input and output of your program. ✔

Extra Credit - Not found

Fundamentals
Place all shared project properties into a Directory.Build.Props file.
Place all shared project items into a Directory.Build.targets file.
nullable reference types is enabled ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ❌
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ❌
and enabled .NET analyzers for both projects ❌
For this assignment, always use Assert.AreEqual<T>() (the generic version) ✔
All of the above should be unit tested ✔
Choose simplicity over complexity ✔

File: Calculate/Program.cs (hunk 1 above)
Comment (MattA41) [reply Calculate/Program.cs:27]: will do

File: Calculate/Calculate.csproj [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+    <ImplicitUsings>enable</ImplicitUsings>
+    <Nullable>enable</Nullable>
+	  <LangVersion>latest</LangVersion>
+	  <EnableNetAnalyzer>true</EnableNetAnalyzer>
+	  <AnalysisLevel>latest</AnalysisLevel>
+	  <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
+	  <EnforceCodeStyleInBuild>true</EnforceCodeStyleInBuild>
```
Comment (BenjaminMichaelis) [inline Calculate/Calculate.csproj:12]: These should all be in a props file

File: Calculate/Calculate.csproj [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+	  <EnforceCodeStyleInBuild>true</EnforceCodeStyleInBuild>
```
Comment (BenjaminMichaelis) [inline Calculate/Calculate.csproj:13]: Duplicate element

File: Calculate/Calculator.cs [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+    {
+        if (two == 0)
+        {
+            throw new ArgumentException("Cant divide by 0");
+        }
```
Comment (BenjaminMichaelis) [inline Calculate/Calculator.cs:37]: Divide by zero by default throws a `System.DivideByZeroException`

File: Calculate/Program.cs [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+namespace Calculate;
+
+public class Program
+{
+    public delegate void DelegateWriteLine(string text);
+    public delegate string? DelegateReadLine();
+
+
+    public DelegateWriteLine WriteLine { get; init; }
+    public DelegateReadLine ReadLine { get; init; }
```
Comment (BenjaminMichaelis) [inline Calculate/Program.cs:13]: This isn't quite right, it would be a delegate by just being inline, you don't need both

File: CalculateTests/CalculateTests.csproj [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+  <PropertyGroup>
+    <TargetFramework>net8.0</TargetFramework>
+    <ImplicitUsings>enable</ImplicitUsings>
+    <Nullable>enable</Nullable>
+	  <LangVersion>latest</LangVersion>
+	  <EnableNetAnalyzer>true</EnableNetAnalyzer>
+	  <AnalysisLevel>latest</AnalysisLevel>
+	  <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
+	  <EnforceCodeStyleInBuild>true</EnforceCodeStyleInBuild>
```
Comment (BenjaminMichaelis) [inline CalculateTests/CalculateTests.csproj:11]: Can be in props file

File: CalculateTests/CalculateTests.csproj [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+	  <EnforceCodeStyleInBuild>true</EnforceCodeStyleInBuild>
```
Comment (BenjaminMichaelis) [inline CalculateTests/CalculateTests.csproj:12]: Duplicate element

File: CalculateTests/CalculatorTest.cs
Comment (BenjaminMichaelis) [inline CalculateTests/CalculatorTest.cs:1]: Using theory would allow you to test more edge cases quickly
